import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/CMoveService.java

```diff
@@ -0,0 +1,240 @@
+          qidoPath = AttributesUtil.attributesToQidoPath(keysCopy,
+              TagUtils.toHexString(Tag.TransferSyntaxUID));
+          log.info("CMove QidoPath: " + qidoPath);
+        } catch (DicomServiceException e) {
+          log.error("CMove QidoPath error");
+          sendErrorResponse(e.getStatus(), e.getMessage(), e.getDataset(), null);
+          return;
+        }
+
+        JSONArray qidoResult;
+        try {
+          qidoResult = dicomWebClient.qidoRs(qidoPath);
+          if (qidoResult == null || qidoResult.length() == 0) {
+            throw new IDicomWebClient.DicomWebException("No instances to move");
+          }
+        } catch (IDicomWebClient.DicomWebException e) {
+          log.error("Failed QidoRs for CMove", e);
+          sendErrorResponse(Status.UnableToCalculateNumberOfMatches, e.getMessage());
+          return;
+        }
+
+        cstoreSender = cstoreSenderFactory.create();
+
+        int successfullInstances = 0;
+        int remainingInstances = qidoResult.length();
+        for (Object instance : qidoResult) {
+          sendPendingResponse(remainingInstances, successfullInstances, failedInstanceUids.size());
+
+          if (canceled) {
+            throw new CancellationException();
+          }
+
+          JSONObject instanceJson = (JSONObject) instance;
+          String studyUid = AttributesUtil.getTagValue(instanceJson,
+              TagUtils.toHexString(Tag.StudyInstanceUID));
+          String seriesUid = AttributesUtil.getTagValue(instanceJson,
+              TagUtils.toHexString(Tag.SeriesInstanceUID));
+          String instanceUid = AttributesUtil.getTagValue(instanceJson,
+              TagUtils.toHexString(Tag.SOPInstanceUID));
+          String classUid = AttributesUtil.getTagValue(instanceJson,
+              TagUtils.toHexString(Tag.SOPClassUID));
+          String tsuid = AttributesUtil.getTagValueOrNull(instanceJson,
+              TagUtils.toHexString(Tag.TransferSyntaxUID));
+          if (tsuid == null) {
+            tsuid = UID.ExplicitVRLittleEndian;
+          }
+
+          try {
+            MonitoringService.addEvent(Event.CMOVE_CSTORE_REQUEST);
+            long bytesSent = cstoreSender.cstore(cstoreTarget, studyUid, seriesUid,
+                instanceUid, classUid, tsuid);
+            successfullInstances++;
+            MonitoringService.addEvent(Event.CMOVE_CSTORE_BYTES, bytesSent);
+          } catch (IDicomWebClient.DicomWebException | IOException e) {
+            MonitoringService.addEvent(Event.CMOVE_CSTORE_ERROR);
+            log.error("Failed CStore within CMove", e);
+            failedInstanceUids.add(instanceUid);
+          }
+
+          remainingInstances--;
+        }
+
+        if (failedInstanceUids.isEmpty()) {
+          as.tryWriteDimseRSP(pc, Commands.mkCMoveRSP(cmd, Status.Success));
+        } else {
+          int status = successfullInstances > 0 ?
+              Status.OneOrMoreFailures : Status.UnableToPerformSubOperations;
```

Comment:
From reading http://dicom.nema.org/medical/dicom/current/output/chtml/part04/sect_C.4.2.3.html#sect_C.4.2.3.1 both seem equally valid.
DCM4che chose to return UnableToPerformSubOperations in same situation https://dcm4che.atlassian.net/browse/DCMEE-30 . So at least I'm not the only one to pick this code.

To me UnableToPerformSubOperations seems to be more clear indication that problem is with sub operations rather than c-move as whole, but I can change to more generic status if you think it would be better.

Reply:
Ok, if that's what dcm4che does then this seems fine as is.